### PR TITLE
threads: restrict number of golang threads if not set

### PR DESCRIPTION
**Pull request by grahamwhaley**
If the max number of golang threads (GOMAXPROCS) is not set in
the environment, then set it ourselves to a small set, but only if
there are more cores on the host than our desired max.
This avoids us taking the default (the number of host cores)
on large-core systems, which can end up with us having a lot of
idle threads generated, consuming host PID space, and thus
artificially restricting how many containers we can run
in parallel.

Fixes: #117

Signed-off-by: Graham Whaley <graham.whaley@intel.com>

### CODE REVIEW

File: proxy.go [hunk 1]
```diff
@@ -352,6 +353,21 @@ func realMain() {
...
 
+	// If GOMAXPROCS has not been set, restrict our thread usage
+	// so we don't grow many idle threads on large core count systems,
+	// which un-necessarily consume host PID space (and thus set an
+	// artificial max limit on the number of concurrent containers we can
+	// run)
+	if os.Getenv("GOMAXPROCS") == "" {
+		// We choose 6 as we want a couple of threads for the runtime (gc etc.)
+		// and couple of threads for our parallel user code copy routines.
+		const maxthreads = 6
```
Comment (jodh-intel) [inline proxy.go]: Might be worth putting this at the top of the file to make it stand out a little more

**Comment by jodh-intel** [comment]
CI is failing due to cyclomatic complexity:

```
proxy.go:352::warning: cyclomatic complexity 17 of function realMain() is high (> 15) (gocyclo)
```

**Comment by grahamwhaley** [comment]
@jodh-intel @marcov both this and the parallel shim PR updated - let's see how they fly.

**Comment by marcov** [comment]
@grahamwhaley nice, lgtm.

File: proxy.go [hunk 2]
```diff
@@ -348,10 +353,25 @@ func handleVersion(showVersion bool) {
 	}
 }
 
+func setThreads() {
+	// If GOMAXPROCS has not been set, restrict our thread usage
+	// so we don't grow many idle threads on large core count systems,
+	// which un-necessarily consume host PID space (and thus set an
+	// artificial max limit on the number of concurrent containers we can
+	// run)
+	if os.Getenv("GOMAXPROCS") == "" {
```
Comment (jodh-intel) [inline proxy.go]: Just a thought - I wonder if it's worth adding a log call showing this value? The complication being that you can only do that after `setupLogger()` has been called. But you could still call `setThreads()` and then add the log call for `GOMAXPROCS` later I guess.

**Comment by grahamwhaley** [comment]
np. If we were doing some more dynamic thread euthanasia or adaptation whilst running, then yes, I'd log it. But this is a one time moderately static startup thing, so I'll not log. Let's see how long it is before I regret that decision ;-)

**Comment by devimc** [comment]
looks g o o d , could you run a round of PnP tests on this PR?

File: proxy.go [hunk 3]
```diff
@@ -30,6 +31,10 @@ import (
 const (
 	proxyName  = "kata-proxy"
 	termSignal = syscall.SIGTERM
+	// Max number of threads the proxy should consume.
+	// We choose 6 as we want a couple of threads for the runtime (gc etc.)
+	// and couple of threads for our parallel user code copy routines.
+	maxThreads = 6
```
Comment (devimc) [inline proxy.go]: uhmmm 6 threads for a POD with N containers (shims)...

Comment (grahamwhaley) [reply proxy.go]: :-) I was hoping somebody might debate the '6'.
(you no doubt know this, but...) So, with 6 threads you can run 6 bits of code in parallel - like, really in parallel. You can have as many goroutines pending or sleeping and scheduled to execute. How many goroutines do we think we actually execute really in parallel, that are not blocked on IO or are that time critical? We are not afaik compute bound....
I'm more than happy to bump the number, but I'd just like to be sure we will actually use those parallel threads in parallel, and not just consume host PIDs for no real reason.

**Comment by jodh-intel** [comment]
Same problem as https://github.com/kata-containers/shim/pull/115 - "Failed to load JSON" - is metrics broken?

**Comment by grahamwhaley** [comment]
A bunch of the 'fringe' CIs failed - I have a feeling there is no good reason why that would happen, so I'm going to re-nudge a full CI run and we'll see where we end up...
/test

**Comment by grahamwhaley** [comment]
Some mighty odd CI fails here - not sure they are related - so, let's try again....

```
Running setup
Running ramdisk test
50d8ba1e861ac62db5db1dd9b0ef8cfc5c2b5c149e71f443b3fa4137a1a826dc
Running teardown
integration/ramdisk/ramdisk.sh: line 63: 15445 Killed                  ( sudo -E DOCKER_RAMDISK=true dockerd -D --add-runtime ${RUNTIME}=${RUNTIME_PATH} 2> /dev/null )
/tmp/jenkins/workspace/kata-containers-proxy-ubuntu-16-04-PR/go/src/github.com/kata-containers/proxy
/dev/fd/63: line 1: html: No such file or directory
/dev/fd/63: line 2: syntax error near unexpected token `<'
/dev/fd/63: line 2: `<head><title>504 Gateway Time-out</title></head>
'
Build step 'Execute shell' marked build as failure
```

/test

**Comment by sboeuf** [comment]
@grahamwhaley looks good, but why are the metrics tests failing?

**Comment by grahamwhaley** [comment]
The metrics build slave is down right now. We were getting too much noise, and I'm nearly done deploying a bare-metal instance. The metrics CI seems to be not registering on new PRs, so not blocking them, but this is an older PR, so had a stale CI attached to it.
Hopefully the metrics CI will be back up in the first half of next week.

**Pull request merged**